src/api/worker/rest/CacheStorageProxy.ts

```diff
@@ -0,0 +1,54 @@
+import {CacheStorage} from "./EntityRestCache"
+import {ProgrammingError} from "../../common/error/ProgrammingError"
+
+export type StorageInitArgs = {persistent: true, databaseKey: Uint8Array, userId: Id} | {persistent: false}
+
+export interface LateInitializedCacheStorage {
+	uninitializedStorage: CacheStorage
+
+	initializeCacheStorage(args: StorageInitArgs): Promise<void>
+}
+
+/**
+ * This is necessary so that we can release offline storage mode without having to rewrite the credentials handling system. Since it's possible that
+ * a desktop user might not use a persistent session, and we won't know until they try to log in, we can only decide what kind of cache storage to use at login
+ * This implementation allows us to avoid modifying too much of the worker public API. Once we make this obsolete, all we will have to do is
+ * remove the initialize parameter from the LoginFacade, and tidy up the WorkerLocator init
+ *
+ * Create a proxy to a cache storage object.
+ * It will be uninitialized, and unusable until {@method LateInitializedCacheStorage.initializeCacheStorage} has been called on the returned object
+ * Once it is initialized, then it is safe to use
+ * @param factory A factory function to get a CacheStorage implementation when initialize is called
+ * @return {LateInitializedCacheStorage} The uninitialized proxy and a function to initialize it
+ */
+export function makeStorageProxy(factory: (args: StorageInitArgs) => Promise<CacheStorage>): LateInitializedCacheStorage {
+	let actualStorage: CacheStorage | null = null
+
+	const proxy = new Proxy({}, {
```

Comment:
oh I thought you wouldn't make an actual `Proxy` but just another impl with deletegation but I guess this also works?

Reply:
Is there a reason it wouldn't work?

Reply:
Idk, proxy gives unexpected result but I think it should be fine